### ISSUE TITLE
Spec the behavior of precision_config in DotGeneralOp

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#755]
See https://github.com/openxla/stablehlo/pull/307#discussion_r1045135311.

**Comment by ghpvnist** [comment]
Here are the results of my investigation:
On the [JAX](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L657-L670) side, this is what the enum values mean:
* DEFAULT: uses `bf16` for computation with accumulation done in at least the precision of the output type, for which `preferred_element_type` (#600) can override, in which case depending on whether the precision is larger or smaller, the accumulation can have more/less precision, respectively.
* HIGH: uses 3 `bf16` passes or `tf32` (8 exponent, 10 mantissa bits) type if available for computation. The accumulation is done in a similar fashion.
* HIGHEST: uses `f32` or `f64` as applicable for computation. The accumulation is done in a similar fashion.

For StableHLO, we currently do not have `preferred_element_type` as an (optional) attribute and rely on the result element type. This means that we always have a defined `preferred_element_type`, so accumulation would always happen using the result element type. Also, we do not have `tf32` or model passes to calculate multiplication by splitting the bits (as this is more hardware specific). Therefore, I propose we use the following to do computation and accumulation based on the `precison_config` values:
* DEFAULT: use `bf16` for computation, accumulating values using the result element type as `preferred_element_type`.
* HIGH: use `f32` for computation, accumulating values using the result element type as `preferred_element_type`.
* HIGHEST: use `f64` for computation, accumulating values using the result element type as `preferred_element_type`.

In other words, the multiplication/sum (intermediate results) is calculated using `bf16`, `f32`, or `f64`, and the resulting sum (final value) is casted to the result element type.

**Comment by ghpvnist** [comment]
(C1) lhs and rhs have the same element type.

Currently, (C1) does not allow lhs and rhs element types to be different. This would prevent specifying distinct preferred element types for lhs and rhs which means that having two config values does not provide extra information since this behavior can be captured with just one config value. This constraint should also be removed while this issue is addressed.

**Comment by zichuan-wei** [comment]
Given that the inputs and outputs of an op already contain information about the precision of the expected input and output, I think it makes not so much sense to have a separate precision config, especially for on-device use cases. And DEFAULT, HIGH, & HIGHEST is kind of vague in terms of what they represent.

**Comment by sdasgup3** [comment]
Thanks @zichuan-wei for the input! 
IMO, precision_config's one use case is to override the precision specified by the input parameters, allowing the choice   of low precision ( and fast) vs  high precision (and slow) computation. I think this param is being used in server GPU and TPUs. 

>> And DEFAULT, HIGH, & HIGHEST are kind of vague in terms of what they represent.
I agree with you.  I recently learned from @reedwm how involving the semantics of these are and I believe fixing the vagueness is a priority for this ticket.

**Comment by sdasgup3** [comment]
AI: `precision_config` and quantization do not go hand in hand and there behavior should be defined when when co-exist (for example, convolution with `precision_config`).
